### DISCUSSION TITLE
Stop updating table heading

### DISCUSSION THREAD

**Discussion by rmitesh** [filamentphp/filament#6802] [Help] [answered]
I have one widget where I have display loop through in the page using categories.

So, one widget and display dynamic tables using the category.
```php
protected function getHeaderWidgets(): array
{
    return [
        Widgets\PlanProjectList::class,
    ];
}
```

And I have created custom function for the set heading in the `PlanProjectList` class

```php
public static function setHeader($string): void
{
    static::$heading = $string;
}

public static function getHeader(): string | null
{
    return static::$heading;
}
```

```blade
@if ($headerWidgets = $this->getVisibleHeaderWidgets())
    <x-filament-support::grid
        :default="$columns['default'] ?? 1"
        :sm="$columns['sm'] ?? null"
        :md="$columns['md'] ?? null"
        :lg="$columns['lg'] ?? ($columns ? (is_array($columns) ? null : $columns) : 2)"
        :xl="$columns['xl'] ?? null"
        :two-xl="$columns['2xl'] ?? null"
        class="filament-widgets-container gap-4 lg:gap-8 mb-6"
    >
    	@foreach( $project_categories as $k => $project_category )
	        @foreach ($headerWidgets as $widget)
		    	@php
		    		$widget::setHeader($project_category);
		    	@endphp

	            @if ($widget::canView())
	                @livewire(\Livewire\Livewire::getAlias($widget), $widgetData, key($k))
	            @endif
	        @endforeach
	    @endforeach
    </x-filament-support::grid>
@endif
```
![image](https://github.com/filamentphp/filament/assets/48554454/77441a1d-a6b9-4d91-9b40-dfee9c84a3a1)

In the `PlanProjectList` `getTableHeaderActions()` I have created custom Table Action called "Add Project"
On click it's open a modal, but table heading is updating to original class name that is "Plan Project List"

![image](https://github.com/filamentphp/filament/assets/48554454/d7d0b92c-86ca-4bf8-bdf7-ea7e2dcb0417)

Is there any way to stop updating table heading?

**Comment by danharrin** [comment]
Livewire doesn't work like that, static variables are not persisted between requests. All you need to do is pass through the heading into `@livewire` as a parameter, using a `$heading` public property to persist it, and then return that from `getTableHeading()`

**Reply by rmitesh** [reply]
@danharrin 
Don't know I did correctly or not, but I have applied following code...

Set public variable 

```php
class Todos extends PageWidget
{
    public static ?string $heading = null;
    // ....
}
```

Page blade file

```blade
@if ($headerWidgets = $this->getVisibleHeaderWidgets())
    <x-filament-support::grid
        :default="$columns['default'] ?? 1"
        :sm="$columns['sm'] ?? null"
        :md="$columns['md'] ?? null"
        :lg="$columns['lg'] ?? ($columns ? (is_array($columns) ? null : $columns) : 2)"
        :xl="$columns['xl'] ?? null"
        :two-xl="$columns['2xl'] ?? null"
        class="filament-widgets-container gap-4 lg:gap-8 mb-6"
    >

        @foreach ($categories as $category)
            @foreach ($headerWidgets as $widget)
                @if ($widget::canView())
                	@php
                		$widget::$heading = $category; // CustomerSetting::$heading = $category
                		$data['heading'] = $category;
                	@endphp
                    @livewire(\Livewire\Livewire::getAlias($widget), $data, key($widget))
                @endif
            @endforeach
        @endforeach
    </x-filament-support::grid>
@endif
```

Also, getting this error when changing to `public static ?string $heading;`
![image](https://github.com/filamentphp/filament/assets/48554454/b81d4457-f4b4-4e0b-ac37-9572a5153e56)

**Reply by danharrin** [reply]
I dont think `$heading` should be static

**Reply by rmitesh** [reply]
Yes, it should be else it's throwing an error

**Reply by rmitesh** [reply]
Does any hooks call internally, when we click on the Header Action?

**Reply by danharrin** [reply]
Does `PageWidget` contain a `$heading` prop or something? That is a custom class

**Reply by danharrin** [reply]
How about using `$tableHeading` instead or a property that wont clash

**Reply by rmitesh** [reply]
It's an alias class of `Filament\Widgets\TableWidget` which I have used as a `use Filament\Widgets\TableWidget as PageWidget;`

and basically I get an references from the this Filament Tricks [The Power of Filament Actions & Widgets](https://filamentphp.com/tricks/the-power-of-filament-actions-widgets)

**Reply by danharrin** [reply]
Try a new `$tableHeading` property that isnt static and just gets returned from `getTableHeading()`

**Reply by rmitesh** [reply, marked as answer]
I can't control my happiness @danharrin !... it works!... 😍😎🥂

Here is how works:
```php
namespace App\Filament\Pages\Widgets;

use Closure;
use Illuminate\Contracts\Support\Htmlable;
use Filament\Widgets\TableWidget as PageWidget;

class Todos extends PageWidget
{
    public $tableHeading;

    protected function getTableHeading(): string | Htmlable | Closure | null
    {
        return $this->tableHeading;
    }
}
```

and in the blade file
```blade
@foreach( $categories as $k => $category )
    @foreach ($headerWidgets as $widget)
    	@php
            // assign value in to tableHeading variable to livewire component
    		$widgetData['tableHeading'] = $category;
    	@endphp

        @if ($widget::canView())
            @livewire(\Livewire\Livewire::getAlias($widget), $widgetData, key($k))
        @endif
    @endforeach
@endforeach
```

**Comment by rmitesh** [comment]
Thank you so much mate, I learn today something new... cheers🥂
